File: exercism/typescript/wordy/wordy.js
var ParsingErrors;
(function (ParsingErrors) {
    ParsingErrors["UNKNOWN_OPERATION"] = "Unknown operation";
    ParsingErrors["SYNTAX_ERROR"] = "Syntax error";
})(ParsingErrors || (ParsingErrors = {}));
function sanitize(command) {
    let removeBegining = command.replaceAll(new RegExp("What is", "g"), "").trim();
    let removeQuestionMark = removeBegining.replaceAll(new RegExp("[?]", "g"), "");
    return removeQuestionMark;
}
export function doOperation(accumulator, num, operation) {
    switch (operation) {
        case "":
            return num;
        case "plus":
            return accumulator + num;
        case "minus":
            return accumulator - num;
        case "multipliedby":
            return accumulator * num;
        case "dividedby":
            return accumulator / num;
        default:
            throw new Error(ParsingErrors.SYNTAX_ERROR);
    }
}
export const answer = (command) => {
    let result = 0;
    let operatorCount = 0;
    let sanitizedCommand = sanitize(command);
    let tokens = sanitizedCommand.split(" ");
    let operation = "";
    if (command.substring(0, 7) !== "What is") {
        throw new Error(ParsingErrors.UNKNOWN_OPERATION);
    }
    if (isNaN(Number(tokens[0])) || sanitizedCommand === "") {
        throw new Error(ParsingErrors.SYNTAX_ERROR);
    }
    for (let pos = 0; pos < tokens.length; pos++) {
        let currentToken = tokens[pos];
        if (isNaN(Number(currentToken))) {
            operation += currentToken;
        }
        else {
            if (operation === "" && pos > 0) {
                throw new Error(ParsingErrors.SYNTAX_ERROR);
            }
            result = doOperation(result, parseInt(currentToken), operation);
            operation = "";
            operatorCount += 1;
        }
    }
    if (operation === "cubed") {
        throw new Error(ParsingErrors.UNKNOWN_OPERATION);
    }
    if (operation !== "" || operatorCount === 0) {
        throw new Error(ParsingErrors.SYNTAX_ERROR);
    }
    return result;
};

File: exercism/python/wordy/wordy.py
def cut_out_prefix_and_by_suffix(question: str) -> str:
    expected_prefix: str = 'What is'
    if question[:len(expected_prefix)] != expected_prefix:
        raise ValueError("unknown operation")
    return question[len(expected_prefix) + 1:].replace('?', '').replace('by', '')


def calculate(result: int, operation: str, number: int) -> int:
    if operation == 'plus':
        return result + number
    elif operation == 'minus':
        return result - number
    elif operation == 'divided':
        return result / number
    elif operation == 'multiplied':
        return result * number
    else:
        raise ValueError("unknown operation")


def answer(question: str) -> int:
    question = cut_out_prefix_and_by_suffix(question)
    question_chunked: [str] = question.split()

    if (len(question_chunked)) < 1:
        raise ValueError("syntax error")

    if question_chunked[0].lstrip('-').isdigit():
        result = int(question_chunked.pop(0))
    else:
        raise ValueError("syntax error")

    operation: str = None
    number: int = None
    for chunk in question_chunked:
        if operation is None:
            if chunk in ['plus', 'minus', 'divided', 'multiplied']:
                operation = chunk
                continue
            else:
                if chunk.lstrip('-').isdigit():
                    raise ValueError("syntax error")
                else:
                    raise ValueError("unknown operation")
        if number is None:
            if chunk.lstrip('-').isdigit():
                number = int(chunk)

        if number and operation:
            result = calculate(result, operation, number)
            operation = None
            number = None
        else:
            raise ValueError("syntax error")

    if operation or number:
        raise ValueError("syntax error")

    return result
